radsat-sk/operation/subsystems/RPdb.c

```diff
@@ -0,0 +1,272 @@
+#define PDB_RESPONSE_LENGTH			(4)
+/** Delay between telemetry read/write operations with the PDB on I2C is 5ms */
+#define PDB_I2C_TELEM_DELAY			(5)
+/** Delay between other read/write operations with the PDB on I2C is 1ms */
+#define PDB_I2C_DELAY				(1)
+
+
+/**
+ * Constant Values for I2C Communication
+ */
+/** Number of sun sensors on the CubeSat */
+#define NUM_SUN_SENSORS		(6)
+
+
+
+
+/***************************************************************************************************
+                                          PRIVATE GLOBALS
+***************************************************************************************************/
+/**
+ *  4 Byte commands to request data for each of the 6 sun-sensors
+ *	located on each face of the CubeSat Solar Arrays.
+ */
+static uint32_t PdbSunSensorCommandBytes[6] = {
+	0x10E11C,     // SA1A - xPos? TODO: Fill out which side is which connection
+	0x10E11D,     // SA1B -
+	0x10E12C,     // SA2A -
+	0x10E12D,     // SA2B -
+	0x10E13C,     // SA3A -
+	0x10E13D      // SA3B -
+};
+
+/**
+ * 4 Byte commands for requesting output voltage readings for each bus and BCR
+ */
+static uint32_t PdbVoltageCommandBytes[4] = {
+		0x10E280,		// Output Voltage of BCR
+		0x10E220,		// Output Voltage of Battery Bus
+		0x10E210,		// Output Voltage of 5V Bus
+		0x10E200		// Output Voltage of 3.3V Bus
+};
+
+/**
+ * 4 Byte commands for requesting output current readings for each bus and BCR
+ */
+static uint32_t PdbCurrentCommandBytes[4] = {
+		0x10E284,		// Output Current of BCR in mA
+		0x10E224,		// Output Current of Battery Bus
+		0x10E214,		// Output Current of 5V Bus
+		0x10E204		// Output Current of 3.3V Bus
+};
+
+
+/***************************************************************************************************
+                                       PRIVATE FUNCTION STUBS
+***************************************************************************************************/
+
+static uint16_t PdbI2cTalk(uint32_t command);
+
+
+/***************************************************************************************************
+                                             PUBLIC API
+***************************************************************************************************/
+/**
+ * Get the Irradiance value from the sun sensor on each face of the CubeSat
+ *
+ * @return A pointer to a SunSensorStatus object containing the converted ADC data for each sun sensor
+ */
+SunSensorStatus getSunSensorData(void) {
+
+	// Create a new SunSensorStatus Structure
+	SunSensorStatus sunData;
+
+	// Create a temporary variable for receiving I2C data
+	uint16_t i2c_received;
+
+	// Create a temporary array for calculated sun data before transferring into the structure sunData
+	float convertedData[NUM_SUN_SENSORS];
+
+	// Send 6 commands to get each of the sun sensor's data
+	for(int i = 0; i < NUM_SUN_SENSORS; i = i + 1){
+		i2c_received = PdbI2cTalk(PdbSunSensorCommandBytes[i]);
```

Comment:
Need an error check for your PdbI2cTalk, or else you will load you buffer with the error code and not the data